NetSDK/Services/Metrics/Classes/InMemoryMetricsLog.cs

```diff
@@ -0,0 +1,229 @@
+        {
+            this.apiClient = apiClient;
+            this.countMetrics = countMetrics ?? new ConcurrentDictionary<string, Counter>();
+            this.timeMetrics = timeMetrics ?? new ConcurrentDictionary<string, ILatencyTracker>();
+            this.gaugeMetrics = gaugeMetrics ?? new ConcurrentDictionary<string, long>();
+            this.maxCountCalls = maxCountCalls;
+            this.maxTimeBetweenCalls = maxTimeBetweenCalls;
+            this.countLastCall = utcNowTimestamp;
+            this.timeLastCall = utcNowTimestamp;
+            this.gaugeLastCall = utcNowTimestamp;
+        }
+
+        public void Count(string counterName, long delta)
+        {
+            if (delta <= 0)
+            {
+                return;
+            }
+
+            if (String.IsNullOrEmpty(counterName))
+            {
+                return;
+            }
+
+            Counter counter;
+
+            if (!countMetrics.TryGetValue(counterName, out counter))
+            {
+                counter = new Counter();
+                countMetrics.TryAdd(counterName, counter);
+            }
+
+            counter.AddDelta(delta);
+
+            var oldLastCall = countLastCall;
+            countLastCall = DateTime.UtcNow;
+            if (counter.GetCount() >= maxCountCalls || ((countLastCall - oldLastCall).TotalMilliseconds > maxTimeBetweenCalls))
+            {
+                SendCountMetrics();
+            }
+
+        }
+
+        public void Time(string operation, long miliseconds)
+        {
+            if (String.IsNullOrEmpty(operation) || miliseconds < 0)
+            {
+                return;
+            }
+
+            ILatencyTracker tracker;
+
+            if (!timeMetrics.TryGetValue(operation, out tracker))
+            {
+                tracker = new BinarySearchLatencyTracker();
+                timeMetrics.TryAdd(operation, tracker);
+            }
+
+            tracker.AddLatencyMillis((int)miliseconds);
+
+            var oldLastCall = timeLastCall;
+            timeLastCall = DateTime.UtcNow;
+            if ((timeLastCall - oldLastCall).TotalMilliseconds > maxTimeBetweenCalls)
+            {
+                SendTimeMetrics();
+            }
+        }
+
+        public void Gauge(string gauge, long value)
+        {
+            if (String.IsNullOrEmpty(gauge) || value < 0)
+            {
+                return;
+            }
+
+            if (!gaugeMetrics.ContainsKey(gauge))
+            {
+                gaugeMetrics.TryAdd(gauge, 0);
+            }
+
+            gaugeMetrics[gauge] = value;
+            gaugeCallCount++;
+
+            var oldLastCall = gaugeLastCall;
+            gaugeLastCall = DateTime.UtcNow;
+            if (gaugeCallCount >= maxCountCalls || (gaugeLastCall - oldLastCall).TotalMilliseconds > maxTimeBetweenCalls)
+            {
+                SendGaugeMetrics();
+            }
+        }
+
+
+        private void SendCountMetrics()
```

Comment:
@jwaimann please add a descriptive comment explaining how this look mechanism works.